File: extract_docs.js
const fs = require('fs');
const zlib = require('zlib');
const path = require('path');

function extractDocxText(filePath) {
  try {
    const buffer = fs.readFileSync(filePath);
    let offset = 0;
    
    while (offset < buffer.length - 30) {
      // Look for local file header signature: PK\x03\x04 (0x50 0x4b 0x03 0x04)
      if (buffer[offset] === 0x50 && buffer[offset+1] === 0x4B && buffer[offset+2] === 0x03 && buffer[offset+3] === 0x04) {
        const compMethod = buffer.readUInt16LE(offset + 8);
        const compSize = buffer.readUInt32LE(offset + 18);
        const fileNameLen = buffer.readUInt16LE(offset + 26);
        const extraFieldLen = buffer.readUInt16LE(offset + 28);
        
        const fileName = buffer.toString('utf8', offset + 30, offset + 30 + fileNameLen);
        const dataOffset = offset + 30 + fileNameLen + extraFieldLen;
        
        if (fileName === 'word/document.xml') {
          const compData = buffer.slice(dataOffset, dataOffset + compSize);
          let xmlText;
          if (compMethod === 8) {
            xmlText = zlib.inflateRawSync(compData).toString('utf8');
          } else if (compMethod === 0) {
            xmlText = compData.toString('utf8');
          } else {
            return `Unsupported compression method: ${compMethod}`;
          }
          
          // Simple regex-based XML parser to strip tags and recover formatting
          const textWithParagraphs = xmlText
            .replace(/<\/w:p>/g, '\n')
            .replace(/<w:tab\/>/g, '\t')
            .replace(/<w:br\/>/g, '\n')
            .replace(/<[^>]+>/g, '');
            
          return textWithParagraphs.replace(/\n\n\n+/g, '\n\n').trim();
        }
        offset += 30 + fileNameLen + extraFieldLen + compSize;
      } else {
        offset++;
      }
    }
    return "Could not find word/document.xml in DOCX zip container.";
  } catch (err) {
    return `Error reading file: ${err.message}`;
  }
}

function main() {
  const currentDir = __dirname;
  console.log(`Scanning workspace root for planning files in: ${currentDir}`);
  
  const files = {
    "ML_Dataset_Planning_Document_EV.docx": "ML_Dataset_Planning_Document_EV.txt",
    "Frontend_UI_UX_Planning_Document_EV.docx": "Frontend_UI_UX_Planning_Document_EV.txt"
  };
  
  for (const [docName, txtName] of Object.entries(files)) {
    const docPath = path.join(currentDir, docName);
    const txtPath = path.join(currentDir, txtName);
    
    if (fs.existsSync(docPath)) {
      console.log(`Found: ${docName} (${fs.statSync(docPath).size} bytes)`);
      console.log(`Extracting text to ${txtName}...`);
      const text = extractDocxText(docPath);
      fs.writeFileSync(txtPath, text, 'utf8');
      console.log(`Saved: ${txtName}`);
    } else {
      console.log(`Missing: ${docName} in workspace root.`);
    }
  }
}

main();

File: extract_docs.py
# -*- coding: utf-8 -*-
"""
Helper script to extract text from local DOCX planning files in the workspace root.
Uses only python standard library modules (zipfile and xml.etree).
Run using: python extract_docs.py
"""

import os
import zipfile
import xml.etree.ElementTree as ET

def extract_text_from_docx(docx_path):
    """Extracts text from a .docx file by reading word/document.xml inside the zip package."""
    try:
        with zipfile.ZipFile(docx_path) as docx:
            xml_content = docx.read('word/document.xml')
            root = ET.fromstring(xml_content)
            
            # Namespace map
            ns = {'w': 'http://schemas.openxmlformats.org/wordprocessingml/2006/main'}
            
            text_runs = []
            for elem in root.iter():
                if elem.tag.endswith('t'):
                    if elem.text:
                        text_runs.append(elem.text)
                elif elem.tag.endswith('p') or elem.tag.endswith('br') or elem.tag.endswith('cr'):
                    text_runs.append('\n')
                elif elem.tag.endswith('tab'):
                    text_runs.append('\t')
                    
            full_text = "".join(text_runs)
            while '\n\n\n' in full_text:
                full_text = full_text.replace('\n\n\n', '\n\n')
            return full_text.strip()
    except Exception as e:
        return f"Error extracting text from {docx_path}: {str(e)}"

def main():
    current_dir = os.path.dirname(os.path.abspath(__file__))
    print(f"Scanning workspace root for planning files in: {current_dir}")
    
    files = {
        "ML_Dataset_Planning_Document_EV.docx": "ML_Dataset_Planning_Document_EV.txt",
        "Frontend_UI_UX_Planning_Document_EV.docx": "Frontend_UI_UX_Planning_Document_EV.txt"
    }
    
    for doc_name, txt_name in files.items():
        doc_path = os.path.join(current_dir, doc_name)
        txt_path = os.path.join(current_dir, txt_name)
        
        if os.path.exists(doc_path):
            print(f"Found: {doc_name} (size: {os.path.getsize(doc_path)} bytes)")
            print(f"Extracting text to {txt_name}...")
            text = extract_text_from_docx(doc_path)
            
            with open(txt_path, "w", encoding="utf-8") as f:
                f.write(text)
            print(f"Saved: {txt_name}")
        else:
            print(f"Missing: {doc_name} in workspace root. Please ensure it is present.")

if __name__ == "__main__":
    main()
